timemachine/fe/single_topology_v3.py

```diff
@@ -0,0 +1,539 @@
+    jks = []
+
+    for dg in dummy_groups_b:
+        dummy_atom = list(dg)[0]
+        root_anchors = identify_root_anchors(bond_idxs_b, core_b, dummy_atom)
+
+        if len(root_anchors) == 0:
+            assert 0, "Disconnected dummy group"
+
+        if len(root_anchors) > 1:
+            # Consider the following situation:
+            # D0.D1
+            # .  .  where (.) is the dummy bond
+            # C0-C1
+            #
+            # One of (C0.D0), (D0.D1), (C1.D1) dummy bonds needs to be broken in order to maintain factorizability.
+            # This is a little arbitrary, but some choices are probably more efficient than others:
+            #
+            # hard     easy
+            # D0.D1    D0 D1
+            #    .     .  .
+            # C0-C1    C0-C1
+            #
+            # the lhs is more difficult because it has significantly more phase space that can be
+            # sampled than the fused case, but it's not super obvious how to best detect this.
+            # So instead, we will pick a random anchor atom. One possible solution later on is to
+            # minimize the number of rotatable bonds?
+            warnings.warn(
+                f"Multiple root anchors {root_anchors} found for dummy group: {dg}, picking a random anchor.",
+                MultipleAnchorWarning,
+            )
+
+        jk = None
+
+        # find an arbitrary but stable angle core atom that is one bond away from the root atom that is
+        # present in mol_a
+        for ra in root_anchors:
+            core_b_nbs = [a.GetIdx() for a in mol_b.GetAtomWithIdx(ra).GetNeighbors() if a.GetIdx() in core_b]
+            for nb in core_b_nbs:
+                # see if this core_bond is present in mol_a
+                # first, look up the idxs of the atoms in mol_a:
+                ra_a, nb_a = core_b_to_a[ra], core_b_to_a[nb]
+                if (ra_a, nb_a) in bond_idxs_a or (nb_a, ra_a) in bond_idxs_a:
+                    jk = ra, nb
+                    break
+
+        if jk is None:
+            warnings.warn("Unable to find stable angle term in mol_a", CoreBondChangeWarning)
+            # pick an arbitrary root_anchor
+            jk = root_anchors[0], None
+
+        jks.append(jk)
+
+    return dummy_groups_b, jks
+
+
+class SingleTopologyV3:
+    def __init__(self, mol_a, mol_b, core, forcefield):
+        """
+        SingleTopology combines two molecules through a common core. The combined mol has
+        atom indices laid out such that mol_a is identically mapped to the combined mol indices.
+        The atoms in the mol_b's R-group is then glued on to resulting molecule.
+
+        Parameters
+        ----------
+        mol_a: ROMol
+            First ligand
+        mol_b: ROMol
+            Second ligand
+        core: np.array (C, 2)
+            Atom mapping from mol_a to to mol_b.
+        ff: ff.Forcefield
+            Forcefield to be used for parameterization.
```

Comment:
nit: parameter name in docstring != parameter name in signature

Reply:
willfix